x/mongo/driver/topology/server.go

```diff
@@ -545,18 +546,35 @@ func (s *Server) update() {
 			continue
 		}
 
-		// Must hold the processErrorLock while updating the server description and clearing the
-		// pool. Not holding the lock leads to possible out-of-order processing of pool.clear() and
-		// pool.ready() calls from concurrent server description updates.
-		s.processErrorLock.Lock()
-		s.updateDescription(desc)
-		if err := desc.LastError; err != nil {
-			// Clear the pool once the description has been updated to Unknown. Pass in a nil service ID to clear
-			// because the monitoring routine only runs for non-load balanced deployments in which servers don't return
-			// IDs.
-			s.pool.clear(err, nil)
+		if isShortcut := func() bool {
+			// Must hold the processErrorLock while updating the server description and clearing the
+			// pool. Not holding the lock leads to possible out-of-order processing of pool.clear() and
+			// pool.ready() calls from concurrent server description updates.
+			s.processErrorLock.Lock()
+			defer s.processErrorLock.Unlock()
+
+			s.updateDescription(desc)
+			if lastError := desc.LastError; lastError != nil {
+				// Retry after the first timeout in case of a FAAS pause as described in GODRIVER-2577.
+				if err := unwrapConnectionError(lastError); err != nil {
+					err, ok := err.(net.Error)
+					if ok && err.Timeout() && timeoutCnt < 1 {
```

Comment:
Should we also check for `context.Canceled` and `context.DeadlineExceeded`? We check for those [elsewhere](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/server.go#L449-L454) in `server.go`.